### ISSUE TITLE
New BioSQL test_taxon_left_right_values test requires internet

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#819]
See #809 and https://github.com/biopython/biopython/commit/6348a037dce08e6f64de3fa4b3e2ee99eb1edacc which can give a failure in ``test_taxon_left_right_values`` as follows:

```
======================================================================
ERROR: test_taxon_left_right_values (common_BioSQL.TaxonomyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/Tests/common_BioSQL.py", line 621, in test_taxon_left_right_values
    self.db.load(self.iterator, True)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/BioSQL/BioSeqDatabase.py", line 807, in load
    db_loader.load_seqrecord(cur_record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/BioSQL/Loader.py", line 57, in load_seqrecord
    bioentry_id = self._load_bioentry_table(record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/BioSQL/Loader.py", line 595, in _load_bioentry_table
    taxon_id = self._get_taxon_id(record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/BioSQL/Loader.py", line 195, in _get_taxon_id
    common_name)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/BioSQL/Loader.py", line 425, in _get_taxon_id_from_ncbi_taxon_id
    db="taxonomy", id=ncbi_taxon_id, retmode="XML")
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/Bio/Entrez/__init__.py", line 180, in efetch
    return _open(cgi, variables, post=post)
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/Bio/Entrez/__init__.py", line 556, in _open
    raise exception
  File "/home_local/buildslave/BuildBot_Biopython/lin3464/build/build/lib.linux-x86_64-3.4/Bio/Entrez/__init__.py", line 554, in _open
    handle = _urlopen(cgi)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 153, in urlopen
    return opener.open(url, data, timeout)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 461, in open
    response = meth(req, response)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 574, in http_response
    'http', request, response, code, msg, hdrs)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 499, in error
    return self._call_chain(*args)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 433, in _call_chain
    result = func(*args)
  File "/home_local/buildslave/lib/python3.4/urllib/request.py", line 582, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 503: Service Temporarily Unavailable

----------------------------------------------------------------------
```

Normally we'd just need to move this into a separate like ``test_BioSQL_online.py`` file which declares it requires the internet, but here it probably means another triple set for the different database backends?

@ctSkennerton Am I right in thinking this test does require the internet in order to fetch NCBI taxonomy information via Entrez?

**Comment by peterjc** [comment]
Another example from the same buildbot slave, with a different version of Python (and also seen with Python 2.6 and 2.7), giving a different exception but failing in the same place.

```
======================================================================
ERROR: test_taxon_left_right_values (common_BioSQL.TaxonomyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/Tests/common_BioSQL.py", line 621, in test_taxon_left_right_values
    self.db.load(self.iterator, True)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/BioSQL/BioSeqDatabase.py", line 807, in load
    db_loader.load_seqrecord(cur_record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/BioSQL/Loader.py", line 57, in load_seqrecord
    bioentry_id = self._load_bioentry_table(record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/BioSQL/Loader.py", line 595, in _load_bioentry_table
    taxon_id = self._get_taxon_id(record)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/BioSQL/Loader.py", line 195, in _get_taxon_id
    common_name)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/BioSQL/Loader.py", line 425, in _get_taxon_id_from_ncbi_taxon_id
    db="taxonomy", id=ncbi_taxon_id, retmode="XML")
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/Bio/Entrez/__init__.py", line 180, in efetch
    return _open(cgi, variables, post=post)
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/Bio/Entrez/__init__.py", line 556, in _open
    raise exception
  File "/home_local/buildslave/BuildBot_Biopython/lin3564/build/build/lib.linux-x86_64-3.5/Bio/Entrez/__init__.py", line 554, in _open
    handle = _urlopen(cgi)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 162, in urlopen
    return opener.open(url, data, timeout)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 471, in open
    response = meth(req, response)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 581, in http_response
    'http', request, response, code, msg, hdrs)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 509, in error
    return self._call_chain(*args)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 443, in _call_chain
    result = func(*args)
  File "/home_local/buildslave/lib/python3.5/urllib/request.py", line 589, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 502: Bad Gateway

----------------------------------------------------------------------
```

**Comment by ctSkennerton** [comment]
You're right the test does require the internet to lookup the taxonomy information when loading in the records and then checks to make sure they have been added correctly to the taxon table. Sorry, I didn't realise that tests requiring the internet needed to be separated. I think you are probably right about new test runners needing to be made for each of the database backends. 

I'm happy to fix my mistake up and open a new pull request.

**Comment by peterjc** [comment]
Its a gotcha - sorry, my fault for not noticing this when I reviewed and merged your work. The auto-loading of taxonomy via Entrez works really nicely most of the time, so the test would probably have passed happily if the NCBI or our network hadn't been busy.

Given how we currently split online tests into separate files (to allow ``run_tests.py --offline``) then making three new online BioSQL test files is probably the best solution. 

However, a more fine-grained approach doing this at the test method level would be nicer, perhaps with decorators... CC @tiagoantao in case he has any bright ideas as he was looking at something related with the test framework today.

The rough idea in my head is defining decorators like ``@online`` or  ``@external_dep`` (or ``@nonfree_dep``) decorators which can be linked to the command line switched to ``run_tests.py`` to decide if that test should be skipped or allowed to fail without considering it an error?

(I've not tried implementing anything, just floating an idea here)

**Comment by peterjc** [comment]
Fixed by #847
